File: exercise/src/coursework/DynamicDeque.java
package coursework;

import java.util.ArrayList;
import java.util.List;

class DynamicDeque {
    private List<Object> deque;

    public DynamicDeque() {
        deque = new ArrayList<>();
    }

    public boolean isEmpty() {
        return deque.isEmpty();
    }

    public void appendLeft(Object x) {
        deque.add(0, x);
    }

    public void append(Object x) {
        deque.add(x);
    }

    public void clear() {
        if (!deque.isEmpty()) {
            deque.clear();
            System.out.println("Deque cleared");
        } else {
            System.out.println("Deque already cleared");
        }
    }

    public DynamicDeque copy() {
        DynamicDeque newDeque = new DynamicDeque();
        newDeque.deque.addAll(deque);
        return newDeque;
    }

    public int count(Object x) {
        int count = 0;
        for (Object element : deque) {
            if (element.equals(x)) {
                count++;
            }
        }
        return count;
    }

    public void insert(int i, Object x) {
        if(0 <= i && i <= deque.size()) {
            deque.add(i, x);
        }
        else {
            throw new IndexOutOfBoundsException("Index out of bounds");
        }
    }

    public void extend(Iterable<?> iterable) {
        for (Object element : iterable) {
            append(element);
        }
    }

    public void extendLeft(Iterable<?> iterable) {
        for (Object element : iterable) {
            appendLeft(element);
        }
    }

    public Object pop() {
        if (!isEmpty()) {
            Object removedElement = deque.remove(deque.size() - 1);
            return "Removed element at position " + deque.size() +": " + removedElement;
        } else {
            throw new IllegalArgumentException("Cannot pop from empty deque");
        }
    }

    public Object popLeft() {
        if (!isEmpty()) {
            Object removedElement = deque.remove(0);
            return "Removed element at position 0: " + removedElement;
        } else {
            throw new IllegalArgumentException("Cannot pop from empty deque");
        }
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Object x : deque) {
            sb.append(x).append(" ");
        }
        return "Deque: " + sb.toString();
    }
}

File: dynamic_deque.py
class DynamicDeque:

    def __init__(self):
        self.deque = []

    def is_empty(self):
        return not bool(self.deque)

    def appendleft(self, x):
        self.deque.insert(0, x)

    def append(self, x):
        self.deque.append(x)

    def clear(self):
        if self.deque:
            self.deque = []
            return "Deque cleared"
        else:
            return "Deque already cleared"
    def copy(self):
        new_deque = DynamicDeque()
        new_deque.deque = self.deque.copy()
        return new_deque

    def count(self, x):
        return f"Item {x} is found {self.deque.count(x)} times"

    def insert(self, i, x):
        self.deque.insert(i, x)

    def extend(self, iterable):
        for element in iterable:
            self.append(element)

    def extendleft(self, iterable):
        for element in iterable:
            self.appendleft(element)


    def pop(self):
        if not self.is_empty():
            return self.deque.pop()
        else:
            raise ValueError("Cannot pop from empty deque")

    def popleft(self):
        if not self.is_empty():
            return self.deque.pop(0)
        else:
            raise ValueError("Cannot pop from empty deque")

    def __str__(self):
        return " ".join(str(x) for x in self.deque)
